### ISSUE TITLE
Aug 11 08:24:26 rock2 Keepalived_vrrp[9775]: Missed 1 messages on CPU 4 Aug 11 08:24:26 rock2 Keepalived_vrrp[9775]: Missed -2 messages on CPU 4 Aug 11 08:24:27 rock2 Keepalived_vrrp[9775]: Missed 1 messages on CPU 4 Aug 11 08:24:27 rock2 Keepalived_vrrp[9775]: Missed 2 messages on CPU 5 Aug 11 08:24:27 rock2 Keepalived_vrrp[9775]: Missed -2 messages on CPU 5 Aug 11 08:24:27 rock2 Keepalived_vrrp[9775]: Missed -2 messages on CPU 5

### ISSUE THREAD

**Issue by golgotow** [acassen/keepalived#1959]
**Describe the issue**
log spam with below message, is this harmless ?
Aug 11 08:24:26 rock2 Keepalived_vrrp[9775]: Missed 1 messages on CPU 4
Aug 11 08:24:26 rock2 Keepalived_vrrp[9775]: Missed -2 messages on CPU 4
Aug 11 08:24:27 rock2 Keepalived_vrrp[9775]: Missed 1 messages on CPU 4
Aug 11 08:24:27 rock2 Keepalived_vrrp[9775]: Missed 2 messages on CPU 5
Aug 11 08:24:27 rock2 Keepalived_vrrp[9775]: Missed -2 messages on CPU 5
Aug 11 08:24:27 rock2 Keepalived_vrrp[9775]: Missed -2 messages on CPU 5

**To Reproduce**
root@rock2:~# lsb_release -a; uname -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 20.04.2 LTS
Release:	20.04
Codename:	focal
Linux rock2 4.4.154-112-rockchip-gfdb18c8bab17 #1 SMP Thu Jan 21 04:50:13 UTC 2021 aarch64 aarch64 aarch64 GNU/Linux

3 node keepalived, standard config 

**Expected behavior**
strange message in log

**Keepalived version**
root@rock2:~# keepalived -v
Keepalived v2.0.19 (10/19,2019)

Copyright(C) 2001-2019 Alexandre Cassen, <acassen@gmail.com>

Built with kernel headers for Linux 5.4.18
Running on Linux 4.4.154-112-rockchip-gfdb18c8bab17 #1 SMP Thu Jan 21 04:50:13 UTC 2021

configure options: --build=aarch64-linux-gnu --prefix=/usr --includedir=${prefix}/include --mandir=${prefix}/share/man --infodir=${prefix}/share/info --sysconfdir=/etc --localstatedir=/var --disable-silent-rules --libdir=${prefix}/lib/aarch64-linux-gnu --runstatedir=/run --disable-maintainer-mode --disable-dependency-tracking --with-kernel-dir=debian/ --enable-snmp --enable-sha1 --enable-snmp-rfcv2 --enable-snmp-rfcv3 --enable-dbus --enable-json --enable-bfd --enable-regex build_alias=aarch64-linux-gnu CFLAGS=-g -O2 -fdebug-prefix-map=/build/keepalived-FYhPQP/keepalived-2.0.19=. -fstack-protector-strong -Wformat -Werror=format-security LDFLAGS=-Wl,-Bsymbolic-functions -Wl,-z,relro CPPFLAGS=-Wdate-time -D_FORTIFY_SOURCE=2

Config options:  NFTABLES LVS REGEX VRRP VRRP_AUTH JSON BFD OLD_CHKSUM_COMPAT FIB_ROUTING SNMP_V3_FOR_V2 SNMP_VRRP SNMP_CHECKER SNMP_RFCV2 SNMP_RFCV3 DBUS

System options:  PIPE2 SIGNALFD INOTIFY_INIT1 VSYSLOG EPOLL_CREATE1 IPV4_DEVCONF IPV6_ADVANCED_API LIBNL3 RTA_ENCAP RTA_EXPIRES RTA_NEWDST RTA_PREF FRA_SUPPRESS_PREFIXLEN FRA_SUPPRESS_IFGROUP FRA_TUN_ID RTAX_CC_ALGO RTAX_QUICKACK RTEXT_FILTER_SKIP_STATS FRA_L3MDEV FRA_UID_RANGE RTAX_FASTOPEN_NO_COOKIE RTA_VIA FRA_OIFNAME FRA_PROTOCOL FRA_IP_PROTO FRA_SPORT_RANGE FRA_DPORT_RANGE RTA_TTL_PROPAGATE IFA_FLAGS IP_MULTICAST_ALL LWTUNNEL_ENCAP_MPLS LWTUNNEL_ENCAP_ILA NET_LINUX_IF_H_COLLISION LIBIPVS_NETLINK IPVS_DEST_ATTR_ADDR_FAMILY IPVS_SYNCD_ATTRIBUTES IPVS_64BIT_STATS IPVS_TUN_TYPE IPVS_TUN_CSUM IPVS_TUN_GRE VRRP_VMAC VRRP_IPVLAN IFLA_LINK_NETNSID CN_PROC SOCK_NONBLOCK SOCK_CLOEXEC O_PATH GLOB_BRACE INET6_ADDR_GEN_MODE VRF SO_MARK SCHED_RT SCHED_RESET_ON_FORK

**Distro (please complete the following information):**
root@rock2:~# lsb_release -a; uname -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 20.04.2 LTS
Release:	20.04
Codename:	focal
Linux rock2 4.4.154-112-rockchip-gfdb18c8bab17 #1 SMP Thu Jan 21 04:50:13 UTC 2021 aarch64 aarch64 aarch64 GNU/Linux

**Details of any containerisation or hosted service (e.g. AWS)**
If keepalived is being run in a container or on a hosted service, provide full details

**Configuration file:**
A full copy of the configuration file, obfuscated if necessary to protect passwords and IP addresses

**Notify and track scripts**
If any notify or track scripts are in use, please provide copies of them

**System Log entries**
Full keepalived system log entries from when keepalived started

**Did keepalived coredump?**
If so, can you please provide a stacktrace from the coredump, using gdb.

**Additional context**

**Comment by pqarmitage** [comment]
What you are seeing isn't expected. It appears that proc connector messages are being delivered out of order, which I have not seen before.

However, you appear to be running a version of keepalived that has been built for a 5.4 series kernel on a 4.4.154 kernel and this isn't supported. keepalived must be built with kernel headers for the kernel version on which it is running (or an earlier kernel version). The kernel version for Ubuntu 20.04 is 5.4, so I am not clear why you are running a much older kernel.

If you can't upgrade to a 5.4 kernel, you will need to build keepalived using kernel headers for the kernel that you are running. It the problem still persists after that then we can look at this further.

Please ensure that you provide a copy of your configuration so that we can test what you are experiencing.

**Comment by golgotow** [comment]
Hello,

Usage of older kernel is forced mostly as hardware is rock pi 4, and they provided this version of kernel for now. 
So you say I shall build from sources keepalived instead of install it via apt as is now ? If yes I will try in near future.

Best regards

**Comment by pqarmitage** [comment]
Yes, you need to build keepalived from source using the kernel headers
that match the kernel you are running.

There will be other programs/utilities, that will expect certain kernel
features to exist that don't, such as the ip command (iproute), nft (for
nftables). Those programs don't build based on what kernel features are
supported by the kernel headers, but later versions of the code expect
certain new kernel features to exist, which of course won't in your case.
I remember managing to set up something in nftables using a version of the
nft program that was for a later version than the running kernel.

Could you please also provide a copy of your keepalived configuration so
that I can try testing it. I don't have a Rock Pi 4 but I do have a
Raspberry Pi 4 running a 64 bit kernel on which I can test it.

**Comment by golgotow** [comment]
those are 3 copy of same config, as I see mostly functions works without problem, only this spam in log 

! Configuration File for keepalived

global_defs {
   notification_email {
     mail@gmail.com
   }
   notification_email_from lb3@example.com
   smtp_server 192.168.1.50
   smtp_connect_timeout 30
}

vrrp_track_process track_proces {
      process proces
      weight 10
}

vrrp_script keepalived_check {
      script "/etc/keepalived/check-proces.sh"
      interval 1
      timeout 5
      rise 3
      fall 3
      weight 10
}

vrrp_instance VI_PROCES_1 {
    state MASTER
    interface eth0
    virtual_router_id 133
    priority 100
    advert_int 1
    authentication {
        auth_type PASS
        auth_pass password
    }
    virtual_ipaddress {
        192.168.1.100/22
    }
    track_script {
       keepalived_check
    }
    notify /etc/keepalived/keepalived_notify.sh
}


/etc/keepalived/check-k3s.sh 
#!/bin/bash
ps -ef | grep -i "usr/local/bin/proces" | grep -v "grep

**Comment by pqarmitage** [comment]
```
Aug 11 08:24:26 rock2 Keepalived_vrrp[9775]: Missed 1 messages on CPU 4
Aug 11 08:24:26 rock2 Keepalived_vrrp[9775]: Missed -2 messages on CPU 4
Aug 11 08:24:27 rock2 Keepalived_vrrp[9775]: Missed 1 messages on CPU 4
Aug 11 08:24:27 rock2 Keepalived_vrrp[9775]: Missed 2 messages on CPU 5
Aug 11 08:24:27 rock2 Keepalived_vrrp[9775]: Missed -2 messages on CPU 5
Aug 11 08:24:27 rock2 Keepalived_vrrp[9775]: Missed -2 messages on CPU 5
```
Messages delivered out of sequence is a problem, since a PROC_EVENT_EXIT may be delivered before the PROC_EVENT_FORK reporting the creation of the process.

I have found a kernel commit - ab8ed951080e - with the title _connector: fix out-of-order cn_proc netlink message delivery_, which describes exactly what you appear to be experiencing. The commit was first included in Linux v4.7, so that explains why you are experiencing out of sequence message delivery.

Regarding the above, considering CPU 4 the following sequence of messages numbers would generate what you saw:
540
542
541
543
which has no missing message but the second and third messages are delivered out of sequence.

However, on CPU 5 it is not possible for there simply to have been a reordering of messages since the sum of the messages missed is not equal to 0 (unless there were also other Missed messages logged).

Since Linux v4.7 was released over 5 years ago I am not inclined to attempt to do anything about this. However, if strong pressure were applied by sufficient people I might think about it since Linux v4.4 is an LTS kernel. I think my first approach then would be to see if the above kernel patch could be backported into the v4.4 kernel, rather than trying to create new code in keepalived.
